### PR TITLE
chore: switch to esm

### PR DESCRIPTION
**Pull request by vasco-santos**
BREAKING CHANGE: built content includes ESM and CJS

Needs: 

- [x] https://github.com/ipfs/aegir/pull/865
- [x] bundle action back in place

### CODE REVIEW

**Comment by vasco-santos** [comment]
This should be ready, just needs a new `aegir`. It would be helpful to have a review @achingbrain

File: package.json [hunk 1]
```diff
@@ -48,15 +25,51 @@
...
+    "./compare": {
+      "import": "./src/compare.js"
+    },
+    "./concat": {
+      "import": "./src/concat.js"
+    },
+    "./equals": {
+      "import": "./src/equals.js"
+    },
+    "./fromString": {
```
Comment (achingbrain) [inline package.json:63]:
```suggestion
    "./from-string": {
```

File: package.json [hunk 2]
```diff
@@ -48,15 +25,51 @@
...
+      "import": "./src/from-string.js"
+    },
+    "./toString": {
```
Comment (achingbrain) [inline package.json:66]:
```suggestion
    "./to-string": {
```

**Review by achingbrain — CHANGES_REQUESTED** [review]

File: tsconfig.json [hunk 3]
```diff
@@ -1,12 +1,15 @@
...
-    "include": [
-      "test",
-      "util",
-      "*.js",
-      ".aegir.js"
-    ]
-}
+  "extends": "aegir/src/config/tsconfig.aegir.json",
+  "compilerOptions": {
+    "declarationDir": "./dist/types",
```
Comment (achingbrain) [inline tsconfig.json:4]:
```suggestion
    "outDir": "./dist/types",
```

File: package.json [hunk 4]
```diff
@@ -48,15 +25,51 @@
...
+      "sourceType": "module"
+    },
     "ignorePatterns": [
       "!.aegir.js"
     ]
   },
+  "typesVersions": {
+    "*": {
+      "*": [
+        "types/*"
```
Comment (achingbrain) [inline package.json:43]:
```suggestion
        "types/src/*"
```

File: package.json [hunk 5]
```diff
@@ -48,15 +25,51 @@
...
+      ],
+      "types/*": [
+        "types/*"
```
Comment (achingbrain) [inline package.json:46]:
```suggestion
        "types/src/*"
```

File: package.json [hunk 6]
```diff
@@ -2,40 +2,17 @@
...
-    "index.js",
-    "to-string.js",
-    "xor.js",
-    "dist/*.ts",
-    "dist/*.map",
-    "dist/*.js",
-    "util/*.js"
-  ],
+  "type": "module",
+  "types": "dist/types/index.d.ts",
```
Comment (achingbrain) [inline package.json:10]:
```suggestion
  "types": "types/src/index.d.ts",
```

**Comment by achingbrain** [comment]
I think all the files need to switch over to using named exports, otherwise this innocuous looking sort of thing:

```js
/**
 * @typedef {import('foo').Bar} Bar
 */

module.exports = function baz () {}
```

Generates:

```ts
export type Bar = import('foo').Bar;
export default function baz(): void;
```

and ipjs says to [not mix default and named exports](https://github.com/mikeal/ipjs#do-not-use-mixed-default-default-and-named-exportsimports).

File: package.json [hunk 7]
```diff
@@ -48,15 +25,51 @@
...
   },
+  "typesVersions": {
+    "*": {
+      "*": [
+        "types/*"
+      ],
+      "types/*": [
+        "types/*"
+      ]
+    }
```
Comment (achingbrain) [inline package.json:48]:
```suggestion
    "*": {
      "*": [
        "types/src/*"
      ]
    }
```

**Comment by achingbrain** [comment]
> I think all the files need to switch over to using named exports

Actually maybe not, we can just remove the `@typedef` and import the type directly in the function's `@param` annotation, this solves that problem.

Still getting some weird `typeof import('path/to/uint8arrays/types/src/from-string') has no call signatures` errors though.

**Comment by achingbrain** [comment]
No, we have to switch to named exports.

`has no call signatures` - this is caused by ts wanting:

```js
const fromString = require('uint8arrays/from-string')
```

to be:

```js
const fromString = require('uint8arrays/from-string').default
```

which compiles types but then fails at runtime in node due to `"require"` in the exports map resolving to the `.cjs` file which doesn't have a `.default` property:

```js
"exports": {
  "browser": "./esm/src/from-string.js",
  "require": "./cjs/src/from-string.js",
  "import": "./esm/src/from-string.js"
}
```

Ignoring that for a second, if we run bundled cjs code in the browser, `"exports"` means the browser gets supplied the esm version, which *does* have a `.default` property so breaks - fixing this would require a PR to ipjs to change [this line](https://github.com/mikeal/ipjs/blob/master/src/package/index.js#L241) to supply the cjs version instead or at least make it an option.  Do we want that?  We want to live in the bundler-free future so probably not.

Ingoring *that* for a second, if we follow my [suggestion above](https://github.com/achingbrain/uint8arrays/pull/22#issuecomment-888058558):

> we can just remove the @typedef and import the type directly in the function's @param annotation

we're breaking compatibility because we previously exported a `SupportedEncodings` type from `to-string` and `from-string` which would be removed, so we'd need a major.

We could ditch cjs, which would require all cjs users to change all instances of `require('uint8arrays/from-string')` to `require('uint8arrays/from-string').default` so we'd need a major.

Instead we can switch to named exports, which will need a major, but then people don't need to use `.default` and we can retain the `SupportedEncodings` type.

As ever, this springs to mind:

image

**Comment by vasco-santos** [comment]
Thanks for the detailed feedback. I agree that the best solution is to make named exports and I will change it

**Comment by vasco-santos** [comment]
types tested on https://github.com/ipfs/js-ipfs/pull/3774

File: .github/workflows/main.yml [hunk 8]
```diff
@@ -17,9 +17,6 @@ jobs:
     - uses: gozala/typescript-error-reporter-action@v1.0.8
     - run: npx aegir build
     - run: npx aegir dep-check
-    - uses: ipfs/aegir/actions/bundle-size@master
```
Comment (vasco-santos) [inline .github/workflows/main.yml:20]: this needs to be added again once aegir ships it as I could not make it use a branch

Comment (achingbrain) [reply .github/workflows/main.yml]: I've restored this but now CI is failing on this step with:

```
Error: Cannot find module '/github/workspace/index.js'
```

File: package.json [hunk 9]
```diff
@@ -48,15 +25,70 @@
...
+    },
+    "./from-string.js": {
+      "import": "./src/from-string.js"
+    },
+    "./to-string.js": {
+      "import": "./src/to-string.js"
+    },
+    "./xor.js": {
+      "import": "./src/xor.js"
+    }
```
Comment (achingbrain) [inline package.json:90]: Why do we need to double up these exports?  Could be quite error prone over time, maybe we could get ipjs to do this for us?

File: .github/workflows/main.yml (hunk 8 above)
Comment (vasco-santos) [reply .github/workflows/main.yml]: oh damn, I will look into it

File: package.json (hunk 9 above)
Comment (vasco-santos) [reply package.json]: We talked about this offline a couple of weeks. Anyway, we do not need this anymore. Removed now

**Review by achingbrain — APPROVED** [review]

**Pull request merged**